### PR TITLE
feat: add go rtdb emulator support

### PR DESCRIPTION
**Pull request by r-LaForge**
### Discussion

This PR hopes to add support for the RTDB Emulator, and is discussed here: https://github.com/firebase/firebase-admin-go/issues/298

There was a previous attempt here: https://github.com/firebase/firebase-admin-go/pull/471

The [Python](https://github.com/firebase/firebase-admin-python/blob/master/firebase_admin/db.py) and [Node](https://github.com/firebase/firebase-admin-node/blob/master/src/database/database.ts) SDK emulation was followed as an example.

### Testing

Unit tests have been added to ensure the emulator config (url + namespace) is set correctly. All others still pass.
I have also adjusted the integration tests to also be able to run against the emulator. 
Simply run:
```
# localhost:9000 simply being the default port.
 FIREBASE_DATABASE_EMULATOR_HOST=localhost:9000 FIREBASE_DATABASE_EMULATOR_NAMESPACE=<your-db> go test ./integration/db
```

Please note the `FIREBASE_DATABASE_EMULATOR_NAMESPACE` is only used to set the database name during tests. 

### API Changes

No public API changes.

RELEASE NOTE: The Realtime Database API now supports running against Firebase emulator suite. Enable the emulator mode by setting the `FIREBASE_DATABASE_EMULATOR_HOST` environment variable to point to the emulator Database endpoint.

### CODE REVIEW

File: db/db.go [hunk 1]
```diff
@@ -126,3 +152,53 @@ func handleRTDBError(resp *internal.Response) error {
...
+	baseURL := strings.Replace(rawEmulatorHostURL, fmt.Sprintf("?%s", parsedEmulatorHost.RawQuery), "", -1)
+	if parsedEmulatorHost.Scheme != "http" {
+		baseURL = fmt.Sprintf("http://%s", baseURL)
+	}
+
+	namespace := parsedEmulatorHost.Query().Get(emulatorNamespaceParam)
+	if namespace == "" {
+		if strings.Contains(rawEmulatorHostURL, ".") {
+			namespace = strings.Split(rawEmulatorHostURL, ".")[0]
+		}
```
Comment (r-LaForge) [inline db/db.go:194]: I observed similar behaviour in the Node sdk: https://github.com/firebase/firebase-admin-node/blob/master/src/database/database.ts#L325

File: db/db.go [hunk 2]
```diff
@@ -126,3 +153,65 @@ func handleRTDBError(resp *internal.Response) error {
...
+	parsedURL, err = url.ParseRequestURI(environmentEmulatorURL)
+	if err != nil {
+		return nil, false, fmt.Errorf("%s: %w", environmentEmulatorURL, ErrInvalidURL)
+	}
+	cfg, err := parseEmulatorHost(environmentEmulatorURL, parsedURL)
+	return cfg, true, err
+}
+
+func parseEmulatorHost(rawEmulatorHostURL string, parsedEmulatorHost *url.URL) (*dbURLConfig, error) {
+	if strings.Contains(rawEmulatorHostURL, "//") {
```
Comment (r-LaForge) [inline db/db.go:194]: I saw this behavior in the Python sdk: https://github.com/firebase/firebase-admin-python/blob/37ecf18d0ad09d7402143f22443e957a57ec7a2a/firebase_admin/db.py#L784

**Comment by r-LaForge** [comment]
I've signed the CLA check and I'm not sure why `google-oss` is showing up as a contributor.
image

File: db/db.go [hunk 3]
```diff
@@ -126,3 +153,65 @@ func handleRTDBError(resp *internal.Response) error {
...
 	return err
 }
+
+// parseURLConfig returns the dbURLConfig for the database
+// dbURL may be either:
+//   - a production url (https://foo-bar.firebaseio.com/)
+//   - an emulator URL (localhost:9000/?ns=foo-bar)
+//
+// The following rules will apply for determining the output:
+//   - If the url has no scheme it will be assumed to be an emulator url and be used.
```
Comment (r-LaForge) [inline db/db.go:163]:
```suggestion
//   - If the url does not use an https scheme it will be assumed to be an emulator url and be used.
```

**Comment by r-LaForge** [comment]
Thank you for kicking off CI
I fixed the lint issue.

**Comment by lahirumaramba** [comment]
> I've signed the CLA check and I'm not sure why `google-oss` is showing up as a contributor. image
> 
> other PRs appear to have this same issue: #511

@r-LaForge I have seen this happen if we switch the base branch from to `dev` from the github UI. Did you by any chance based your initial changes on the main branch? If that's the case could you try rebasing to `dev` locally and push the changes?

**Comment by r-LaForge** [comment]
@lahirumaramba Rebase shows branch is up to date `Current branch feature/go-rtdb-emulator is up to date.`.
Branching directly from `dev` and creating another PR with the specific file changes results in the same outcome: https://github.com/firebase/firebase-admin-go/pull/519

**Review by yuchenshi — APPROVED** [review]

**Review by christhompsongoogle — APPROVED** [review]

**Review by lahirumaramba — APPROVED** [review]

**Comment by lahirumaramba** [comment]
Thank you so much @r-LaForge for your contribution! We will include this change in the next release of the Admin Go SDK.

**Pull request merged**